velox/experimental/cudf/exec/ToCudf.cpp

```diff
@@ -0,0 +1,219 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/experimental/cudf/exec/CudfConversion.h"
+#include "velox/experimental/cudf/exec/CudfOrderBy.h"
+#include "velox/experimental/cudf/exec/ToCudf.h"
+#include "velox/experimental/cudf/exec/Utilities.h"
+
+#include "velox/exec/Driver.h"
+#include "velox/exec/FilterProject.h"
+#include "velox/exec/HashAggregation.h"
+#include "velox/exec/HashBuild.h"
+#include "velox/exec/HashProbe.h"
+#include "velox/exec/Operator.h"
+#include "velox/exec/OrderBy.h"
+
+#include <cudf/detail/nvtx/ranges.hpp>
+
+#include <cuda.h>
+
+#include <iostream>
+
+namespace facebook::velox::cudf_velox {
+
+namespace {
+
+template <class... Deriveds, class Base>
+bool isAnyOf(const Base* p) {
+  return ((dynamic_cast<const Deriveds*>(p) != nullptr) || ...);
+}
+
+} // namespace
+
+bool CompileState::compile() {
+  auto operators = driver_.operators();
+  auto& nodes = planNodes_;
+
+  // Make sure operator states are initialized.  We will need to inspect some of
+  // them during the transformation.
+  driver_.initializeOperators();
+
+  bool replacementsMade = false;
+  auto ctx = driver_.driverCtx();
+
+  // Get plan node by id lookup.
+  auto getPlanNode = [&](const core::PlanNodeId& id) {
+    auto it =
+        std::find_if(nodes.cbegin(), nodes.cend(), [&id](const auto& node) {
+          return node->id() == id;
+        });
+    VELOX_CHECK(it != nodes.end());
+    return *it;
+  };
+
+  auto isSupportedGpuOperator = [](const exec::Operator* op) {
+    return isAnyOf<exec::OrderBy>(op);
+  };
+
+  std::vector<bool> isSupportedGpuOperators(operators.size());
+  std::transform(
+      operators.begin(),
+      operators.end(),
+      isSupportedGpuOperators.begin(),
+      isSupportedGpuOperator);
+
+  auto acceptsGpuInput = [](const exec::Operator* op) {
+    return isAnyOf<exec::OrderBy>(op);
+  };
+
+  auto producesGpuOutput = [](const exec::Operator* op) {
+    return isAnyOf<exec::OrderBy>(op);
+  };
+
+  int32_t operatorsOffset = 0;
+  for (int32_t operatorIndex = 0; operatorIndex < operators.size();
+       ++operatorIndex) {
+    std::vector<std::unique_ptr<exec::Operator>> replaceOp;
+
+    exec::Operator* oper = operators[operatorIndex];
+    auto replacingOperatorIndex = operatorIndex + operatorsOffset;
+    VELOX_CHECK(oper);
+
+    bool const kPreviousOperatorIsNotGpu =
+        (operatorIndex > 0 and !isSupportedGpuOperators[operatorIndex - 1]);
+    bool const kNextOperatorIsNotGpu =
+        (operatorIndex < operators.size() - 1 and
+         !isSupportedGpuOperators[operatorIndex + 1]);
+
+    auto id = oper->operatorId();
+    if (kPreviousOperatorIsNotGpu and acceptsGpuInput(oper)) {
```

Comment:
Before that, we can use PlanNode->toString to see the node, but how can we see the really plan operator executed?

Reply:
I'm not sure I understand the question. Could you please elaborate with an example?
Are you asking if there's a way to find out whether the operator used is a cudf based one or velox cpu based one? We add the prefix "Cudf" to the operator name like so https://github.com/devavret/velox/blob/ab399f22c8393e3e9a99a60c46fd94e8f2edbb88/velox/experimental/cudf/exec/CudfOrderBy.cpp#L38. Ideally this should reflect in the operator.toString() method, unless the latter is overloaded.

Reply:
We only send a PlanNode to Task, and print the PlanNode->toString by the runtime setting, but now, the operator is hidden by PlanNode, only if we debug step by step or add the code to print operators, we can see the difference. Can we add a flag like FLAGS_velox_time_allocations to print the operator information? We can set this flag by runtime setting which is friendly to debug.

Reply:
We used to have an environment variable for printing out pipelines before and after operator replacement. Should I bring that back as a gflag?

Reply:
Does this work? https://github.com/facebookincubator/velox/pull/12735/commits/7570d56d06d70f7211de382161b7ae6d7f19be14

Reply:
Number of new operators: 5 -> Number of operators: 5
Looks good.